File: Python/is_this_a_bst.py
# Solution using inorder traversal
''' This solution does nto submit on hackerrank '''


class Solution:
    def inorder(self, root,l):
        if(root==None):
            return
        self.inorder(root.left,l)
        l.append(root.val)
        self.inorder(root.right,l)
        
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        l=[]
        self.inorder(root,l)
        for i in range(len(l)-1):
            if(l[i]>=l[i+1]):
                return False
        else:
            return True
   

# Solution using property of BST
''' This solution submits of hackerrank '''

def isBST(root, min_v, max_v):
    if(root == None):
        return True
    if(root.data < min_v or root.data > max_v):
        return False
    return (isBST(root.left, min_v, root.data-1) and isBST(root.right, root.data+1, max_v))

def check_binary_search_tree_(root):
    INT_MIN = float('-inf')
    INT_MAX = float('inf')
    return isBST(root, INT_MIN, INT_MAX)

File: Java/January/jan14/is_this_a_bst.java
package Java.jan14;

import java.util.*;
class Solution
{   // Using inorder traversal
    public static void inorder(Node root, LinkedList<Integer> list)
    {
        if(root==null)
            return;
        inorder(root.left, list);
        list.add(root.data);
        inorder(root.right, list);
    }

    boolean checkBST(Node root) {
        LinkedList<Integer> list = new LinkedList<>();
        Tree.inorder(root, list);
        for(int i=0; i<list.size()-1; i++)
        {
            if(list.get(i)>=list.get(i+1))
                return false;
        }
        return true;
    }
    //Inorder solution end


    //Using property of BST
    public boolean checkBST(Node root, int min, int max)
    {
        if(root==null)
            return true;
        if(root.data < min || root.data > max)
            return false;
        return (checkBST(root.left, min, root.data-1) && checkBST(root.right, root.data+1, max));

    }

    public boolean checkBST(Node root) {
        return checkBST(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }
    //BST Property solution end.
}


